rfcs/0001-fobnail-architecture-change.md

```diff
@@ -0,0 +1,207 @@
+4. Platform Owner responds with AIK challenge and AIK object ID.
+5. Attester performs Credential Activation using the provided challenge and
+   sends results to Platform Owner with EK object ID and AIK object ID.
+6. Platform Owner verifies the results of Credential Activation. If verification
+   is successful, AIK is marked as trusted, and the Platform Owner creates
+   Provisioning Context and sends its ID (PC ID) to the client.
+7. Attester signs and sends its metadata to the Fobnail Token, together with PC
+   ID. Platform Owner verifies metadata signature against AIK bound to specified
+   Provisioning Context.
+8. Platform Owner verifies metadata and returns status. Metadata is bound to the
+   specified Provisioning Context.
+9. Attester sends RIMs.
+10. Platform Owner verifies RIMs and returns status. RIMs are bound to the
+    Provisioning Context.
+11. At this point, additional calls may be issued to send more data. This
+    feature is left for future provisioning extensions.
+12. Attester finalizes platform provisioning by sending a special request.
+13. Platform Owner verifies all the data bound to the Provisioning Context (AIK
+    and RIMs currently)
+14. Platform Owner sends its certificate chain to Fobnail Token.
+15. Fobnail Token verifies certificate chain and returns Provisioning Context
+    ID.
+16. Platform Owner signs the Platform metadata with private matching the leaf
+    certificate (sent as part of the chain in the previous step) and sends
+    signed Platform metadata to Fobnail Token.
+17. Fobnail Token verifies metadata signature and sends status.
+18. Platform Owner sends signed RIMs to Fobnail Token.
+19. Fobnail Token verifies RIMs signature and responds with status.
+20. Platform Owner sends (signed) policy to Fobnail Token.
+21. Fobnail Token verifies policy and responds with status.
+22. At this point Platform Owner may issue additional calls to send more data.
+    This feature is left for future provisioning extensions.
+23. Platform Owner completes provisioning by sending a special request.
+24. Fobnail verifies all the data bound to the Provisioning Context (AIK and
+    RIMs currently), saves it to flash, and sends status.
+25. Platform Owner sends the final status to the provisioned Platform.
+26. Platform is provisioned now.
+
+Remote attestation doesn't differ in anything from the Attester's point of view,
+and the protocol is exactly the safe as if Attester has spoken directly to the
+Fobnail Token.
+
+## Platform attestation
+
+Since Fobnail Token is a server, the attestation must be requested by Attester.
+After plugging Fobnail Token into USB, the attestation must start and complete
+within a defined amount of time. Otherwise, Fobnail Token will consider the
+Platform untrustworthy.
+
+1. Attester starts attestation by sending signed platform metadata to Fobnail
+   Token.
+2. Fobnail Token verifies metadata and returns Attestation Context ID together
+   with PCR selection and nonce.
```

Comment:
@krystian-hebel I was thinking about just incrementing by 1 each newly generated ID, but we may as well use RNG.

Reply:
Incremented counter is easier, with RNG you have to keep track of duplicates. Then again, if IDs are predictable, replay attacks are possible. Maybe use incremented IDs, but start from random initial value?

Reply:
Wouldn't that reduce ramdomness?
I was thinking of ID as a 4 byte number only for purpose of tracking temporary objects. We are using 32 bytes for nonce and object ID has to be encoded as string (CoAP requirements), later this has to be passed as part of request URI, using 32 bytes for object ID would increase message sizes instead of reducing them.